File: leetcode/sliding_window/NumSubarrayProductLessThanK.java
/*
713. Subarray Product Less Than K
Medium
Topics
Companies
Hint
Given an array of integers nums and an integer k, return the number of contiguous subarrays where the product of all the elements in the subarray is strictly less than k
*/

class NumSubarrayProductLessThanK {
    /*
    Approach: Sliding window, 2 pointers template
    Time: O(n) Space:O(1)
    */
    private static int numSubarrayProductLessThanK(int[] nums, int k) {
        int start = 0, currProd = 1, count = 0;

        for(int end = 0; end<nums.length; end++) {
        	currProd *= nums[end];

        	while (start<=end && currProd >= k) {
        		currProd /= nums[start++];
        	}

        	count += end-start+1;
        }
        return count;
    }

    public static void main(String[] args) {
    	int[] nums = {10,5,2,6};
    	int k = 100;
    	System.out.println(numSubarrayProductLessThanK(nums, k));
    }
}

File: leetcode/sliding_window/caterpillar/num_subarray_product_less_than_k.py
"""
713. Subarray Product Less Than K
Solved
Medium
Topics
Companies
Hint
Given an array of integers nums and an integer k, return the number of contiguous subarrays where the product of all the elements in the subarray is strictly less than k.
"""


class NumSubarrayProductLessThanK:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        numSub = 0
        winProd = 1
        left = 0

        for right in range(len(nums)):
            winProd *= nums[right]

            while left <= right and winProd >= k:
                winProd = winProd // nums[left]
                left += 1

            # Intuition 1:
            #   Num of subarrays satisfying given condition = len of subarray (r - l + 1)
            # Intuition 2:
            #   This is the num of subarrays ending at right pointer
            numSub += right - left + 1

        return numSub
